Port Cifar Augmented pipeline and its nodes

shivaram opened the pull request:
Fixes #242

shivaram commented:
cc @ericmjonas @Vaishaal 

Some notes:
1. This lacks any unit tests right now. I wanted to get this out early to get some feedback on interfaces. I will work on unit tests soon

2. To do augmentation I was forced to use `FunctionNode` -- I don't think we have a better way to do this ? Also not being able to chain function nodes in the middle of a pipeline is a pain.

3. It'll be good to see if some of these APIs like `RandomFlipper` can be generalized like @ericmjonas mentioned in the other thread.

etrain commented on src/main/scala/evaluation/AugmentedExamplesEvaluator.scala:
```diff
@@ -0,0 +1,69 @@
+import org.apache.spark.rdd.RDD
+import nodes.util.MaxClassifier
+
+object AugmentedExamplesEvaluator extends Serializable {
```
Instead of `AugmentedExamplesEvaluator` this seems like it could just as easily be called `EnsembleEvaluator` or something. 

In the traditional ensemble setting, a hypothetical `EnsembleEstimator` could be an `Estimator[T,Seq[V]]` or if people wanted to construct ensembles themselves, they could use a `Pipeline.gather` to coallate the results of running several models on an input data item.

Here we have a case that's slightly different, but I think the basic idea still holds.

shivaram replied:
This is now tracked in #250 and #249

etrain commented on src/main/scala/evaluation/AugmentedExamplesEvaluator.scala:
```diff
@@ -0,0 +1,69 @@
+    preds.reduce(_ + _) :/ preds.size.toDouble
+  }
+
+  // * Let s(k) be the ordering of patch k.
```
Maybe change this to a scaladoc?

shivaram replied:
Done

etrain commented on src/main/scala/evaluation/AugmentedExamplesEvaluator.scala:
```diff
@@ -0,0 +1,69 @@
+    ranks.reduceLeft(_ + _)
+  }
+
+  def apply[T: ClassTag](
```
Given my comment above - I wonder if this names RDD is the right thing to do, or whether the grouping to a `Array[DenseVector[Double]]` makes sense in another node? What I'm struggling with a little bit conceptually here is the logical pipeline that we want to describe. Conceptually, featurization is like a map from an Image to a Matrix of features. Those features are concatenated and a model is fit. At test time there is no concatenation and we need an averaging policy. 

This is just one way to look at it, though - and I guess it is worth sketching out variants of the pipeline on a whiteboard or something.

shivaram replied:
Same as above - This is now tracked in #250 and #249

etrain commented on src/main/scala/nodes/images/CenterCornerPatcher.scala:
```diff
@@ -0,0 +1,48 @@
+  * @param horizontalFlips if horizontal flips of patches should also be returned
+  * @return patches of size patchSizeX x patchSizeY
+  */
+case class CenterCornerPatcher(
```
Can this be accomplished with a correctly configured `Windower`,  `HorizontalFlips`, and a `Gather` - not saying we wouldn't also want a node called this that is the composition of these things - just trying to avoid having a node that feels pretty special purpose.

shivaram replied:
Well we need a function node to do this `flatMap`. And I don't think the windower is appropriate here is it gets all the matches with size patchSize (we can throw out the rest of them but that seems like same amount of code with wasted work). 

I think having `ImageUtils.crop` and `ImageUtils.flipHorizontal` make this function pretty readable, but yeah this node does seem overly-specialized

etrain commented on src/main/scala/nodes/images/Pooler.scala:
```diff
@@ -67,3 +67,21 @@ class Pooler(
+
+object Pooler {
+
+  def sumPooler(in: DenseVector[Double]): Double = {
```
What's wrong with `sum(in)`?

shivaram replied:
Its slower - and constantly showed up in `jstack` while I was running experiments. To get a sense of how much slower here is a unscientific benchmark I ran on my laptop
```
var sum=0.0; var sumTime=0.0; 
for (i <- 0 until 10000) { 
  val a = DenseVector.rand(4096);
  val begin = System.nanoTime(); 
  sum += sumPooler(a);  // Replace this with a.sum for the Breeze one
  sumTime += (System.nanoTime() - begin)/1000.0 
};
```
Avg sumTime for breeze: 163.9452 microseconds
Avg sumTime for while loop: 21.4573 microseconds

At length=8192
Avg sumTime for breeze: 328.2782 microseconds
Avg sumTime for while loop: 40.1276 microseconds

At length=81920
Avg sumTime for breeze: 3351.0207 microseconds
Avg sumTime for while loop: 412.5542 microseconds

So this 8x overhead seems to be consistent as we increase the vector length.

etrain replied:
Hmm.. ".sum" is deprecated, and on *my* laptop - `sum(x)` (well, with an appropriate change to "var sum=0.0" above), goes 2x faster than this while loop.

shivaram replied:
I see the same thing - Not sure why the while loop might be slower, but I feel this isn't worth investigating. Changed it to just use `sum(_)` now

etrain commented on src/main/scala/nodes/images/Pooler.scala:
```diff
@@ -67,3 +67,21 @@ class Pooler(
+  }
+
+  def avgPooler(in: DenseVector[Double]): Double = {
+    sumPooler(in) / in.length
```
Similarly, why not `breeze.stats.mean`?

shivaram replied:
Removed now

etrain commented on src/main/scala/nodes/images/RandomPatcher.scala:
```diff
@@ -0,0 +1,47 @@
+import pipelines.FunctionNode
+
+/**
+  * Extract random square patches from an image
```
We should probably mention that patches are selected *uniformly* at random.

shivaram replied:
Done

etrain commented:
Alright - code-wise, this mostly looks good. The higher level question is how we want to represent this pipeline conceptually. Is it a flatmap with some key attached to the features that we can later group on (this is messy because it assumes labels and features have identical partitioning, etc.) or would we rather think of this in more general terms with some kind of ensembling. I think this is one of those situations where a whiteboard and an example pipeline might be useful, then we can decide if we have sufficient tools to represent that thing or not.

shivaram commented:
Cool - We can chat about this on Monday. I don't think we have the right machinery to do this kind of augment-ensembles cleanly yet, but we can try to use this pipeline to design things

Vaishaal commented on src/main/scala/nodes/images/RandomFlipper.scala:
```diff
@@ -0,0 +1,26 @@
+  */
+case class RandomFlipper(flipChance: Double, seed: Long = 12334L) extends Transformer[Image, Image] {
+
+  val rnd = new scala.util.Random(seed)
```
Apply scala.util.Random isn't serializable...

```
Caused by: java.io.NotSerializableException: scala.util.Random
Serialization stack:
        - object not serializable (class: scala.util.Random, value: scala.util.Random@39652a30)
        - field (class: nodes.images.RandomPatcher, name: rnd, type: class scala.util.Random)
        - object (class nodes.images.RandomPatcher, <function1>)
        - field (class: nodes.images.RandomPatcher$$anonfun$apply$1, name: $outer, type: class nodes.images.RandomPatcher)
        - object (class nodes.images.RandomPatcher$$anonfun$apply$1, <function1>)
        at org.apache.spark.serializer.SerializationDebugger$.improveException(SerializationDebugger.scala:40)
        at org.apache.spark.serializer.JavaSerializationStream.writeObject(JavaSerializer.scala:47)
        at org.apache.spark.serializer.JavaSerializerInstance.serialize(JavaSerializer.scala:84)
        at org.apache.spark.util.ClosureCleaner$.ensureSerializable(ClosureCleaner.scala:301)
        ... 22 more
16/03/13 23:33:37 ERROR ErrorMonitor: AssociationError [akka.tcp://sparkDriver@192.168.10.51:48870] <- [ak
```

shivaram replied:
Uh - thats a pain. Using `java.util.Random` now

Vaishaal commented on src/main/scala/nodes/images/CenterCornerPatcher.scala:
```diff
@@ -0,0 +1,48 @@
+    val yDim = in.metadata.yDim
+
+    val startXs = Array(0, xDim-patchSizeX, 0, xDim-patchSizeX, (xDim-patchSizeX)/2)
+    val startYs = Array(0, 0, yDim-patchSizeY, (yDim-patchSizeY)/2)
```
I think these two lists should be the same length

shivaram replied:
Nice catch. Fixed now

Vaishaal commented on src/main/scala/nodes/images/RandomPatcher.scala:
```diff
@@ -0,0 +1,47 @@
+    patchSizeY: Int,
+    seed: Long = 12334L) extends FunctionNode[RDD[Image], RDD[Image]] {
+
+  val rnd = new scala.util.Random(seed)
```
Same deal about serializability...

shivaram replied:
Done

Vaishaal commented on src/main/scala/pipelines/images/cifar/RandomPatchCifarAugmented.scala:
```diff
@@ -0,0 +1,158 @@
+        trainImages))
+
+    val unscaledFeaturizer = 
+      new Convolver(filters, augmentPatchSize, augmentPatchSize, numChannels, Some(whitener), true)
```
Do you really want to convolve with 24 x 24 patches?

shivaram replied:
That argument is the imageSize as documented in https://github.com/amplab/keystone/blob/master/src/main/scala/nodes/images/Convolver.scala#L22
The patchSize comes from the filters (hmm we are calling both the 6x6 things and the 24x24 things as patches ?)

Vaishaal replied:
yeah that was a point of confusion for me this weekend.

etrain replied:
Hmm.. I wonder if there's a better word to use for the larger "patch" - something like "subImage" or something?

shivaram replied:
I've now just called it image - i.e. `augmentedImgSize` and `numRandomImagesAugmented` -- So other than the nodes the only way `patch` is used is to refer to the convolver patches. Let me know if this looks ok

Vaishaal commented on src/main/scala/pipelines/images/cifar/RandomPatchCifarAugmented.scala:
```diff
@@ -0,0 +1,158 @@
+        ((unnormFilters(::, *) / (twoNorms + 1e-10)) * whitener.whitener.t, whitener)
+    }
+
+    val trainImagesAugmented = RandomFlipper(flipChance).apply(
```
Mostly for my own curiosity..why do we augment train images differently than test?

shivaram replied:
We did this because the original cuda-convnet https://code.google.com/p/cuda-convnet/ code did it this way

shivaram replied:
FWIW this is the link to the writeup https://code.google.com/p/cuda-convnet/wiki/TrainingNet#Training_on_image_translations -- Significant bits are

> ...This will cause the net to train on random 24x24 patches...
After training is done... compute the test error as averaged over 10 patches of the original images (4 corner patches, center patch, and their horizontal reflections). This will produce better results than testing on the center patch alone

shivaram commented:
I've fixed all the comments other the discussion about ensembling. I also introduced a new class `RandomImageTransformer` instead of the `RandomFlipper` per @ericmjonas 's request.

etrain commented on src/main/scala/pipelines/images/cifar/RandomPatchCifarAugmented.scala:
```diff
@@ -0,0 +1,158 @@
+        ((unnormFilters(::, *) / (twoNorms + 1e-10)) * whitener.whitener.t, whitener)
+    }
+
+    val trainImagesAugmented = RandomImageTransformer(flipChance, ImageUtils.flipHorizontal).apply(
```
is this written this way because `FunctionNode` isn't a first-class citizen?

shivaram replied:
Yeah - you can't chain anything after function nodes right now

shivaram commented:
Alright added unit tests for the two patcher nodes and one for `flipHorizontal`. I'm going to run the entire pipeline on the local cluster to make sure we get the same results we got for the paper. @etrain Let me know if you have any other comments on this.

etrain commented on src/main/scala/pipelines/images/cifar/RandomPatchCifarAugmented.scala:
```diff
@@ -0,0 +1,158 @@
+    val trainImages = ImageExtractor(trainData)
+
+    val patchExtractor = new Windower(conf.patchSteps, conf.patchSize)
+      .andThen(ImageVectorizer.apply)
```
style nit - can we move the `andThen` to the end of the previous line and remove the spurious `.`, `(`, `)` and `apply` here? I know you're copy/pasting old code, but I like that style better.

shivaram replied:
Done

shivaram commented:
Other than the performance problem with repeated convolutions, the test error was fine. I got 
```
16/03/21 01:05:04 INFO RandomPatchCifarAugmented: Test error is: 0.1833
```
from using 5120 filters (i.e. 40k features)

etrain commented on src/main/scala/pipelines/images/cifar/RandomPatchCifarAugmented.scala:
```diff
@@ -0,0 +1,157 @@
+
+    val unscaledFeaturizer = 
+      new Convolver(filters, augmentPatchSize, augmentPatchSize, numChannels, Some(whitener), true)
+        .andThen(SymmetricRectifier(alpha=conf.alpha))
```
The next ten or so lines have the old `.andThen(` syntax still.

shivaram replied:
Fixed all the `andThen`s now

shivaram commented:
@etrain the latest commit https://github.com/amplab/keystone/pull/247/commits/202317e2ee045f12373c00a18b12541602575b9f updates the other CIFAR / Mnist pipeline to also use the style described in #258 -- Tests pass, but it would still be good to get another pair of eyes on this commit

etrain commented on src/main/scala/pipelines/images/cifar/RandomCifar.scala:
```diff
@@ -41,18 +47,10 @@ object RandomCifar extends Serializable with Logging {
+        (LinearMapEstimator(conf.lambda), trainImages, trainLabels)
 
-    val predictionPipeline = featurizer andThen model andThen MaxClassifier
+    val predictionPipeline = featurizer andThen MaxClassifier
```
can we add the `MaxClassfier` to the end of what we're calling `featurizer` and change that name to `predictionPipeline`?

shivaram replied:
Yep - done

etrain commented on src/main/scala/pipelines/images/cifar/RandomPatchCifar.scala:
```diff
@@ -50,26 +56,16 @@ object RandomPatchCifar extends Serializable with Logging {
-    val model = new BlockLeastSquaresEstimator(4096, 1, conf.lambda.getOrElse(0.0)).fit(trainFeatures, trainLabels)
-
-    val predictionPipeline = featurizer andThen model andThen MaxClassifier andThen new Cacher[Int]
+    val predictionPipeline = featurizer andThen MaxClassifier andThen new Cacher[Int]
```
Same thing here?

shivaram replied:
done

etrain commented on src/main/scala/pipelines/images/cifar/LinearPixels.scala:
```diff
@@ -23,23 +23,20 @@ object LinearPixels extends Logging {
-    // The final prediction pipeline is the composition of our featurizer and our model.
-    // Since we end up using the results of the prediction twice, we'll add a caching node.
-    val predictionPipeline = featurizer andThen model andThen MaxClassifier
+    val predictionPipeline = featurizer andThen MaxClassifier
```
And similarly here.

shivaram replied:
Done

etrain commented on src/main/scala/pipelines/images/mnist/MnistRandomFFT.scala:
```diff
@@ -43,9 +43,10 @@ object MnistRandomFFT extends Serializable with Logging {
+        MaxClassifier
 
     // Train the model
     val model = pipeline
```
this seems spurious?

shivaram replied:
removed all uses of `model` now

etrain commented:
This LGTM, merging. Thanks @shivaram !
